File: neetcode-150/stack/min-stack/solution.py
"""
Time Complexity: O(1) for all operations
Space Complexity: O(n) - Two stacks
"""
class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if not self.stack:
            return
        
        val = self.stack.pop()
        if val == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

# Alternative approach using single stack with tuples
class MinStackSingle:
    def __init__(self):
        self.stack = []

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append((val, val))
        else:
            min_val = min(val, self.stack[-1][1])
            self.stack.append((val, min_val))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]

# Using deque for better performance
from collections import deque

class MinStackDeque:
    def __init__(self):
        self.stack = deque()
        self.min_stack = deque()

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if not self.stack:
            return
        
        val = self.stack.pop()
        if val == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

# More concise version
class MinStackConcise:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if self.stack.pop() == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

# Using dataclass for cleaner code
from dataclasses import dataclass

@dataclass
class StackElement:
    value: int
    min_value: int

class MinStackDataClass:
    def __init__(self):
        self.stack = []

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append(StackElement(val, val))
        else:
            min_val = min(val, self.stack[-1].min_value)
            self.stack.append(StackElement(val, min_val))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1].value

    def getMin(self) -> int:
        return self.stack[-1].min_value

File: neetcode-150/stack/min-stack/solution.js
/**
 * Time Complexity: O(1) for all operations
 * Space Complexity: O(n) - Two stacks
 */
var MinStack = function() {
    this.stack = [];
    this.minStack = [];
};

MinStack.prototype.push = function(val) {
    this.stack.push(val);
    if (this.minStack.length === 0 || val <= this.minStack[this.minStack.length - 1]) {
        this.minStack.push(val);
    }
};

MinStack.prototype.pop = function() {
    if (this.stack.length === 0) return;
    
    const val = this.stack.pop();
    if (val === this.minStack[this.minStack.length - 1]) {
        this.minStack.pop();
    }
};

MinStack.prototype.top = function() {
    return this.stack[this.stack.length - 1];
};

MinStack.prototype.getMin = function() {
    return this.minStack[this.minStack.length - 1];
};

// Alternative approach using single stack with pairs
var MinStackSingle = function() {
    this.stack = [];
};

MinStackSingle.prototype.push = function(val) {
    if (this.stack.length === 0) {
        this.stack.push([val, val]);
    } else {
        const min = Math.min(val, this.stack[this.stack.length - 1][1]);
        this.stack.push([val, min]);
    }
};

MinStackSingle.prototype.pop = function() {
    this.stack.pop();
};

MinStackSingle.prototype.top = function() {
    return this.stack[this.stack.length - 1][0];
};

MinStackSingle.prototype.getMin = function() {
    return this.stack[this.stack.length - 1][1];
};

// ES6 Class syntax
class MinStackES6 {
    constructor() {
        this.stack = [];
        this.minStack = [];
    }
    
    push(val) {
        this.stack.push(val);
        if (this.minStack.length === 0 || val <= this.minStack[this.minStack.length - 1]) {
            this.minStack.push(val);
        }
    }
    
    pop() {
        if (this.stack.length === 0) return;
        
        const val = this.stack.pop();
        if (val === this.minStack[this.minStack.length - 1]) {
            this.minStack.pop();
        }
    }
    
    top() {
        return this.stack[this.stack.length - 1];
    }
    
    getMin() {
        return this.minStack[this.minStack.length - 1];
    }
}

// More concise version
var MinStackConcise = function() {
    this.stack = [];
    this.minStack = [];
};

MinStackConcise.prototype.push = function(val) {
    this.stack.push(val);
    if (!this.minStack.length || val <= this.minStack[this.minStack.length - 1]) {
        this.minStack.push(val);
    }
};

MinStackConcise.prototype.pop = function() {
    if (this.stack.pop() === this.minStack[this.minStack.length - 1]) {
        this.minStack.pop();
    }
};

MinStackConcise.prototype.top = function() {
    return this.stack[this.stack.length - 1];
};

MinStackConcise.prototype.getMin = function() {
    return this.minStack[this.minStack.length - 1];
};
